waspc/src/Wasp/Analyzer/Parser/Parser.hs

```diff
@@ -0,0 +1,309 @@
+  ( -- * AST to CST conversion
+
+    -- | This module takes @["SyntaxNode"]@ produced by 'Wasp.Analyzer.Parser.ConcreteParser'
+    -- and converts it into an abstract syntax tree.
+    parseStatements,
+    parseExpression,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.State.Strict (StateT, evalStateT, gets, modify)
+import Data.Maybe (catMaybes)
+import Wasp.Analyzer.Parser.AST (AST, Expr, ExtImportName, Identifier, Stmt)
+import qualified Wasp.Analyzer.Parser.AST as AST
+import Wasp.Analyzer.Parser.ConcreteParser.CST (SyntaxKind, SyntaxNode (SyntaxNode))
+import qualified Wasp.Analyzer.Parser.ConcreteParser.CST as S
+import Wasp.Analyzer.Parser.Ctx (Ctx (Ctx), WithCtx (WithCtx), ctxFromRgn)
+import Wasp.Analyzer.Parser.ParseError
+import Wasp.Analyzer.Parser.SourcePosition (SourcePosition (SourcePosition))
+import Wasp.Analyzer.Parser.SourceRegion (SourceRegion (SourceRegion))
+import qualified Wasp.Analyzer.Parser.Token as T
+
+data ParseState = ParseState
+  { pstateLine :: !Int,
+    pstateColumn :: !Int,
+    pstateRemainingSource :: String
+  }
+
+pstatePos :: ParseState -> SourcePosition
+pstatePos s = SourcePosition (pstateLine s) (pstateColumn s)
+
+type ParserM a = StateT ParseState (Except ParseError) a
+
+-- | An operation that turns some syntax nodes into a value and outputs the
+-- remaining nodes.
+type Action a = [SyntaxNode] -> ParserM (a, [SyntaxNode])
+
+-- | @parseStatements sourceString syntax@ tries to convert a concrete syntax
+-- tree into an AST.
+parseStatements :: String -> [SyntaxNode] -> Either ParseError AST
+parseStatements source syntax = runExcept $ evalStateT (coerceProgram syntax) initialState
+  where
+    initialState =
+      ParseState
+        { pstateLine = 1,
+          pstateColumn = 1,
+          pstateRemainingSource = source
+        }
+
+-- | @parseExpression sourceString syntax@ tries to convert a concrete syntax
+-- tree into an AST representing a single expression. Currently, it allows extra
+-- syntax after the single expression.
+parseExpression :: String -> [SyntaxNode] -> Either ParseError Expr
+parseExpression source syntax = case runExcept $ evalStateT (coerceExpr syntax) initialState of
+  Left err -> Left err
+  Right (WithCtx _ expr, _) -> Right expr
+  where
+    initialState =
+      ParseState
+        { pstateLine = 1,
+          pstateColumn = 1,
+          pstateRemainingSource = source
+        }
+
+-- | Try to turn CST into top-level AST.
+coerceProgram :: [SyntaxNode] -> ParserM AST
+coerceProgram [] = return $ AST.AST []
+coerceProgram (SyntaxNode S.Program _ children : _) = AST.AST . catMaybes <$> mapM coerceStmt children
+coerceProgram (SyntaxNode k w _ : ns)
+  | S.syntaxKindIsTrivia k = advance w >> coerceProgram ns
+  | otherwise = failParse $ UnexpectedNode k "in root"
+
+-- | Try to turn CST into Stmt AST. Returns @Nothing@ if the given node is a
```

Comment:
We have to make sure to add the offset of each node we look at. So when we encounter a trivia node, we have to make sure to add its width to the offset. This is sort of mentioned in my response to your comment at the top of this module with what the "contract" is for functions with this type.

Reply:
Could we document this somewhere? As you can see, since we don't have much time left, I am choosing to over-document stuff instead of trying to refactor it / polish it.